libraries/AP_Math/control.cpp

```diff
@@ -179,69 +143,75 @@ void shape_accel_xy(const Vector3f& accel_input, Vector3f& accel,
  The function alters the variable accel to follow a jerk limited kinematic path to vel_input and accel_input
  The accel_max limit can be removed by setting it to zero.
 */
-void shape_vel_accel(const float vel_input1, const float accel_input,
+void shape_vel_accel(const float vel_input, const float accel_input,
                      const float vel, float& accel,
                      const float vel_min, const float vel_max,
                      const float accel_min, const float accel_max,
-                     const float tc, const float dt)
+                     const float jerk_max, const float dt, bool limit_total_accel)
 {
-    // sanity check tc
-    if (!is_positive(tc)) {
+    // sanity check accel_max
+    if (!(is_negative(accel_min) && is_positive(accel_max))) {
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
         return;
     }
 
     // Calculate time constants and limits to ensure stable operation
-    const float KPa = 1.0 / tc;
-
-    // we are changing vel_input, but don't want the change in the caller
-    float vel_input = vel_input1;
-
-    // limit velocity to vel_max
-    if (is_negative(vel_min) && is_positive(vel_max)){
-        vel_input = constrain_float(vel_input, vel_min, vel_max);
-    }
+    const float KPa = jerk_max / accel_max;
 
     // velocity error to be corrected
     float vel_error = vel_input - vel;
 
     // acceleration to correct velocity
     float accel_target = vel_error * KPa;
 
+    // constrain correction acceleration from accel_min to accel_max
+    accel_target = constrain_float(accel_target, accel_min, accel_max);
+
     // velocity correction with input velocity
     accel_target += accel_input;
 
-    shape_accel(accel_target, accel, accel_min, accel_max, tc, dt);
+    // constrain total acceleration from accel_min to accel_max
+    if (limit_total_accel) {
+        accel_target = constrain_float(accel_target, accel_min, accel_max);
+    }
+
+    shape_accel(accel_target, accel, jerk_max, dt);
 }
 
 // 2D version
 void shape_vel_accel_xy(const Vector2f &vel_input1, const Vector2f& accel_input,
                         const Vector2f& vel, Vector2f& accel,
-                        const float vel_max, const float accel_max, const float tc, const float dt)
+                        const float vel_max, const float accel_max,
```

Comment:
let's remove the "const" infront of the "float" for these.

Reply:
It looks like we're not using the vel_max argument anymore.. perhaps it can be removed.